File: dashboard/src/types/work-order.ts
// Work Order types
export type WorkOrderStatus =
  | 'draft'
  | 'assigned'
  | 'in_repair'
  | 'qa'
  | 'completed'
  | 'approved';

export type RepairLocation = 'service_shop' | 'on_site';

export interface WorkOrder {
  id: string;
  incidentId: string;
  tenantId: string;
  schoolId: string;
  deviceId: string;
  schoolName: string;
  contactName: string;
  contactPhone: string;
  contactEmail: string;
  deviceSerial: string;
  deviceAssetTag: string;
  deviceModelId: string;
  deviceMake: string;
  deviceModel: string;
  deviceCategory: string;
  status: WorkOrderStatus;
  serviceShopId: string;
  assignedStaffId: string;
  repairLocation: RepairLocation;
  assignedTo: string;
  taskType: string;
  programId: string;
  phaseId: string;
  onsiteContactId: string;
  approvalStatus: string;
  costEstimateCents: number;
  notes: string;
  // Rework tracking
  reworkCount: number;
  lastReworkAt?: string;
  lastReworkReason: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateWorkOrderRequest {
  incidentId?: string;
  deviceId: string;
  taskType: string;
  serviceShopId?: string;
  assignedStaffId?: string;
  repairLocation?: RepairLocation;
  assignedTo?: string;
  costEstimateCents?: number;
  notes?: string;
}

// Work Order Operations
export interface WorkOrderSchedule {
  id: string;
  tenantId: string;
  schoolId: string;
  workOrderId: string;
  phaseId: string;
  scheduledStart: string | null;
  scheduledEnd: string | null;
  timezone: string;
  notes: string;
  createdByUserId: string;
  createdAt: string;
}

export type DeliverableStatus = 'pending' | 'submitted' | 'approved' | 'rejected';

export interface WorkOrderDeliverable {
  id: string;
  tenantId: string;
  schoolId: string;
  workOrderId: string;
  phaseId: string;
  title: string;
  description: string;
  status: DeliverableStatus;
  evidenceAttachmentId: string;
  submittedByUserId: string;
  submittedAt: string | null;
  reviewedByUserId: string;
  reviewedAt: string | null;
  reviewNotes: string;
  createdAt: string;
  updatedAt: string;
}

export type ApprovalStatus = 'pending' | 'approved' | 'rejected';

export interface WorkOrderApproval {
  id: string;
  tenantId: string;
  schoolId: string;
  workOrderId: string;
  phaseId: string;
  approvalType: string;
  requestedByUserId: string;
  requestedAt: string;
  status: ApprovalStatus;
  decidedByUserId: string;
  decidedAt: string | null;
  decisionNotes: string;
}

// BOM types
export interface WorkOrderPart {
  id: string;
  tenantId: string;
  schoolId: string;
  workOrderId: string;
  serviceShopId: string;
  partId: string;
  partName: string;
  partPuk: string;
  partCategory: string;
  deviceModelId: string;
  isCompatible: boolean;
  qtyPlanned: number;
  qtyUsed: number;
  createdAt: string;
  updatedAt: string;
}

// Rejection/Rework types
export type RejectionCategory = 'quality' | 'incomplete' | 'wrong_parts' | 'safety' | 'other';

export interface WorkOrderReworkHistory {
  id: string;
  tenantId: string;
  schoolId: string;
  workOrderId: string;
  fromStatus: WorkOrderStatus;
  toStatus: WorkOrderStatus;
  rejectionReason: string;
  rejectionCategory: RejectionCategory;
  rejectedByUserId: string;
  rejectedByName: string;
  reworkSequence: number;
  createdAt: string;
}

export interface RejectWorkOrderRequest {
  targetStatus: WorkOrderStatus;
  reason: string;
  category: RejectionCategory;
}

export interface RejectWorkOrderResponse {
  workOrder: WorkOrder;
  reworkHistory: WorkOrderReworkHistory;
}

// Update types
export interface UpdateWorkOrderRequest {
  assignedStaffId?: string;
  serviceShopId?: string;
  costEstimateCents?: number;
  notes?: string;
  repairLocation?: RepairLocation;
  onsiteContactId?: string;
}

export interface UpdateWorkOrderResponse {
  workOrder: WorkOrder;
  updatedFields: string[];
  previousValues?: Record<string, unknown>;
}

// Bulk operation types
export interface BulkStatusUpdateRequest {
  workOrderIds: string[];
  status: WorkOrderStatus;
}

export interface BulkAssignmentRequest {
  workOrderIds: string[];
  assignedStaffId?: string;
  serviceShopId?: string;
}

export interface BulkApprovalRequest {
  workOrderIds: string[];
  decision: 'approved' | 'rejected';
  notes?: string;
}

export interface BulkOperationError {
  id: string;
  message: string;
  code?: string;
}

export interface BulkOperationResult {
  operationId: string;
  succeeded: string[];
  failed: BulkOperationError[];
  totalCount: number;
  successCount: number;
  failureCount: number;
}

File: services/ims-api/internal/models/work_order.go
package models

import "time"

// WorkOrderStatus represents the status of a work order.
type WorkOrderStatus string

const (
	WorkOrderDraft     WorkOrderStatus = "draft"
	WorkOrderAssigned  WorkOrderStatus = "assigned"
	WorkOrderInRepair  WorkOrderStatus = "in_repair"
	WorkOrderQA        WorkOrderStatus = "qa"
	WorkOrderCompleted WorkOrderStatus = "completed"
	WorkOrderApproved  WorkOrderStatus = "approved"
)

// RepairLocation represents where the repair is performed.
type RepairLocation string

const (
	RepairLocationServiceShop RepairLocation = "service_shop"
	RepairLocationOnSite      RepairLocation = "on_site"
)

// WorkOrder represents a repair work order.
type WorkOrder struct {
	ID         string `json:"id"`
	IncidentID string `json:"incidentId"`
	TenantID   string `json:"tenantId"`
	SchoolID   string `json:"schoolId"`
	DeviceID   string `json:"deviceId"`

	// Denormalized lookup fields (from SSOT snapshots)
	SchoolName     string `json:"schoolName"`
	ContactName    string `json:"contactName"`
	ContactPhone   string `json:"contactPhone"`
	ContactEmail   string `json:"contactEmail"`
	DeviceSerial   string `json:"deviceSerial"`
	DeviceAssetTag string `json:"deviceAssetTag"`
	DeviceModelID  string `json:"deviceModelId"`
	DeviceMake     string `json:"deviceMake"`
	DeviceModel    string `json:"deviceModel"`
	DeviceCategory string `json:"deviceCategory"`

	Status            WorkOrderStatus `json:"status"`
	ServiceShopID     string          `json:"serviceShopId"`
	AssignedStaffID   string          `json:"assignedStaffId"`
	RepairLocation    RepairLocation  `json:"repairLocation"`
	AssignedTo        string          `json:"assignedTo"`
	TaskType          string          `json:"taskType"`
	ProjectID         string          `json:"projectId"`
	PhaseID           string          `json:"phaseId"`
	OnsiteContactID   string          `json:"onsiteContactId"`
	ApprovalStatus    string          `json:"approvalStatus"`
	CostEstimateCents int64           `json:"costEstimateCents"`
	Notes             string          `json:"notes"`

	// Project-originated work orders
	CreatedFromProject bool   `json:"createdFromProject"`
	CreatedByUserID    string `json:"createdByUserId"`
	CreatedByUserName  string `json:"createdByUserName"`

	// Rework tracking
	ReworkCount      int        `json:"reworkCount"`
	LastReworkAt     *time.Time `json:"lastReworkAt,omitempty"`
	LastReworkReason string     `json:"lastReworkReason"`

	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

// WorkOrderSchedule represents a scheduled work order session.
type WorkOrderSchedule struct {
	ID              string     `json:"id"`
	TenantID        string     `json:"tenantId"`
	SchoolID        string     `json:"schoolId"`
	WorkOrderID     string     `json:"workOrderId"`
	PhaseID         string     `json:"phaseId"`
	ScheduledStart  *time.Time `json:"scheduledStart"`
	ScheduledEnd    *time.Time `json:"scheduledEnd"`
	Timezone        string     `json:"timezone"`
	Notes           string     `json:"notes"`
	CreatedByUserID string     `json:"createdByUserId"`
	CreatedAt       time.Time  `json:"createdAt"`
}

// DeliverableStatus represents the status of a deliverable.
type DeliverableStatus string

const (
	DeliverablePending   DeliverableStatus = "pending"
	DeliverableSubmitted DeliverableStatus = "submitted"
	DeliverableApproved  DeliverableStatus = "approved"
	DeliverableRejected  DeliverableStatus = "rejected"
)

// WorkOrderDeliverable represents a deliverable for a work order.
type WorkOrderDeliverable struct {
	ID                   string            `json:"id"`
	TenantID             string            `json:"tenantId"`
	SchoolID             string            `json:"schoolId"`
	WorkOrderID          string            `json:"workOrderId"`
	PhaseID              string            `json:"phaseId"`
	Title                string            `json:"title"`
	Description          string            `json:"description"`
	Status               DeliverableStatus `json:"status"`
	EvidenceAttachmentID string            `json:"evidenceAttachmentId"`
	SubmittedByUserID    string            `json:"submittedByUserId"`
	SubmittedAt          *time.Time        `json:"submittedAt"`
	ReviewedByUserID     string            `json:"reviewedByUserId"`
	ReviewedAt           *time.Time        `json:"reviewedAt"`
	ReviewNotes          string            `json:"reviewNotes"`
	CreatedAt            time.Time         `json:"createdAt"`
	UpdatedAt            time.Time         `json:"updatedAt"`
}
